pkg/connmux/connmux.go

```diff
@@ -0,0 +1,335 @@
+// The root listener Accept() method is overriden
+// The multiplexed servers have access to the Close() and Address()
+// methods on the root listener.
+type muxListener struct {
+	net.Listener
+	connc chan net.Conn
+	donec chan struct{}
+}
+
+var _ net.Listener = (*muxListener)(nil)
+
+func (l muxListener) Accept() (net.Conn, error) {
+	select {
+	case c, ok := <-l.connc:
+		if !ok {
+			return nil, net.ErrClosed
+		}
+		return c, nil
+	case <-l.donec:
+		return nil, net.ErrClosed
+	}
+}
+
+// HTTPListener returns a net.Listener that will receive http requests
+func (c *ConnMux) HTTPListener() net.Listener {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.http == nil {
+		c.http = &muxListener{c.root, make(chan net.Conn), c.donec}
+	}
+	return c.http
+}
+
+// GRPCListener returns a net.Listener that will receive grpc requests
+func (c *ConnMux) GRPCListener() net.Listener {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.grpc == nil {
+		c.grpc = &muxListener{c.root, make(chan net.Conn), c.donec}
+	}
+	return c.grpc
+}
+
+// bufferedConn allows to peek in the buffer of the connection
+// without advancing the reader.
+type bufferedConn struct {
+	r *bufio.Reader
+	net.Conn
+}
+
+func newBufferedConn(c net.Conn) bufferedConn {
+	return bufferedConn{bufio.NewReader(c), c}
+}
+
+func (b bufferedConn) Peek(n int) ([]byte, error) {
+	return b.r.Peek(n)
+}
+
+func (b bufferedConn) Read(p []byte) (int, error) {
+	return b.r.Read(p)
+}
+
+func isHTTP2Connection(r io.Reader) bool {
+	// check the http2 client preface
+	buf := make([]byte, len(http2.ClientPreface))
+	n, err := r.Read(buf)
+	if err != nil || n != len(http2.ClientPreface) {
+		return false
+	}
+	return bytes.Equal(buf, []byte(http2.ClientPreface))
+}
+
+func isGRPCConnection(lg *zap.Logger, r io.Reader) bool {
+	done := false
+	isGRPC := false
+	// check if len of the settings frame is 0 or the headers with the "content-type"
+	// indicates is an grpc connection.
+	framer := http2.NewFramer(io.Discard, r)
+	// use the default value for the maxDynamixTable size
+	// https://pkg.go.dev/golang.org/x/net/http2
+	// "If zero, the default value of 4096 is used."
+	hdec := hpack.NewDecoder(4096, func(hf hpack.HeaderField) {
+		// match headers names or values based on https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md
+		if strings.Contains(hf.Name, "grpc") {
+			isGRPC = true
+			done = true
+			lg.Debug("found grpc header", zap.String("header", hf.Name))
+			return
+		}
+		if hf.Name == "content-type" {
+			isGRPC = strings.Contains(hf.Value, "application/grpc")
+			lg.Debug("found content-type header", zap.String("content-type", hf.Value))
+		}
+		done = true
+	})
+	for {
+		f, err := framer.ReadFrame()
+		if err != nil {
+			break
+		}
+		switch f := f.(type) {
+		case *http2.SettingsFrame:
+			// Observed behavior is that etcd GRPC clients sends an empty setting frame
+			// and block waiting for an answer.
+			if f.Length == 0 {
+				isGRPC = true
```

Comment:
What is some grpc client (e.g. the one that doesn't exist today) doesn't send empty settings, rather:
* sends some non-empty settings
* waits for response
* only then sends headers with content-type grpc

It will block on "waits for" before it sends headers.

Reply:
Don't think we care to much about potential future behavior changes in grpc. At least based on my discussion with @dfawley. Still we should verify other officially supported etcd clients like https://github.com/etcd-io/jetcd

Reply:
IIUC, this is not a defined behavior, and is likely to match non-gRPC HTTP/2 clients as well.

In fact, depending on the client's settings, there will be data in there:

https://github.com/grpc/grpc-go/blob/c018273e5353ca8d4106a9d8917f904983b9dbeb/internal/transport/http2_client.go#L419-L431

The handshake we use is the standard HTTP/2 handshake: client sends preface bytes, server sends SETTINGS (which client waits for and ACKs), then client will use the connection for RPCs (which will be the earliest you can tell this is a gRPC client vs. some other HTTP/2 client).

Reply:
basically confirming that is very brittle to implement anything passive that is able to differentiate between http2 and grpc. .... that matches my observation during this exercise ... and since there is no defined behavior we can't guarantee any future compatibility

Reply:
let me add the settings handle to the mux and we revisit this approach